サインアップ機能の実装

daisakuhazui opened the pull request:
# WHAT
deviseを利用してユーザーのサインアップ機能を実装しました。

# WHY
サービスを利用するユーザーは一番最初にサインアップを行います。
開発の順序も、ユーザーエクスペリエンスと同じ順序で行うほうがやりやすいため。

Commit:
不要なコメントアウトの削除

自分で書いたメモ用のコメントアウトが埋もれていたため、不要なコメントアウトを削除しました。
（#後からdefaultオプションを追加する）

Commit:
static_pagesコントローラーの作成

sign_up画面からのサインアップ後、home画面にリダイレクトする仕様にすべく作成しました。（リダイレクト先が存在しないとエラーになるため）

ルーティングも一緒に行えるのでーミナルに、rails g controller static_pages home questionと入力しました。

Commit:
sign_up画面に入力フォームを追加

devise導入時に自動的に生成されたsign_up画面にはalias_nameがないため追加しました。
また、入力フォームの一番上に配置したalias_nameの項目にのみautofocus: trueを設定しています。

emailの項目にあったautofocus: trueは不要なので削除しました。

Commit:
ストロングパラメータの追加

今回、deviseは最新版のdevise4を使用しています。
deviseが配布されているGitHubによればdevise4からストロングパラメータの追加方法が変更されたとのことなので、そちらを参考に記述しました。

https://github.com/plataformatec/devise

Commit:
ルーティングの追加

sing_inまたはsign_upしたしたユーザーはhome画面にリダイレクトをかけます。そのため、rootパスをhome#indexからstatic_pages#homeに変更しました。

homeアクション（home画面を表示）とquestionアクション（友達のquestionを表示）は、static_pagesコントローラを作成したため自動的に生成。

ArakawaShintaro commented on app/assets/javascripts/static_pages.coffee:
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
このファイルは使用していないのであれば削除しましょう
また、自動生成されないように設定しておきましょう

daisakuhazui replied:
了解しました！

ArakawaShintaro commented on app/assets/stylesheets/static_pages.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the StaticPages controller here.
```
こちらも同様です

daisakuhazui replied:
了解しました！

ArakawaShintaro commented on app/helpers/static_pages_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module StaticPagesHelper
```
こちらも使用していないのであればいりませんね
自動生成されないようにしておくといいかもです

daisakuhazui replied:
了解しました！

ArakawaShintaro commented on app/controllers/static_pages_controller.rb:
```diff
@@ -0,0 +1,7 @@
+class StaticPagesController < ApplicationController
```
今回は、サインアップ後のページを作ったと思うのですが
コントローラーが必要な理由はなんでしょうか？

ArakawaShintaro replied:
サインアップ後に、renderとかで直接staticのviewを呼ぶだけでいいと思ったのですが。

ArakawaShintaro replied:
また、homeとquestionのそれぞれの役割もわかりづらいです

daisakuhazui replied:
確かに直接viewを呼ぶだけで良さそうですね。
復習しつつ調べてみます。

また、homeとquestionアクションの役割もぱっと見てわからないですね・・・。
半年後に誰が見たとしても役割がわかるものに変更します。

ArakawaShintaro replied:
> 確かに直接viewを呼ぶだけで良さそうですね。
復習しつつ調べてみます。

了解です！

> また、homeとquestionアクションの役割もぱっと見てわからないですね・・・。

いや、今回でいうとサインアップ後の画面なのでそれがhome画面であればそのままでもいいと思います。
questionがなぜあるかは、不明ですが...
いずれにせよ、controllerは無くなると思うので気にしなくていいかと。

@daisakuhazui

daisakuhazui replied:
これ、controllerって不要なんでしょうか。

home画面で「ログインしている場合は質問が並んだタイムラインを表示する」という実装を考えています。

（questionは、ユーザーがあらかじめ登録していた興味があるジャンルに関する質問をタイムライン形式に表示させようとしていました）

ArakawaShintaro replied:
今の状態で、サインアップしたらどこに遷移していますか？

daisakuhazui replied:
現状、サインアップ後はhome画面に遷移しています。

ArakawaShintaro replied:
そのホーム画面は、Userのshowページとは異なるのですか！？
極力Railsの7つのアクションにそうもの以外は作らない方がいいですね〜
カリキュラム内にhomeというアクションがあったなら、あれはあてにしない方がいいですm(__)m

また、questionも同様です
questionに関してはサインアップ機能と関係ありますかね？

daisakuhazui replied:
何となく理解できてきました。書き直していきますね！

ArakawaShintaro commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -3,9 +3,20 @@
+    <%= f.label :alias_name %><br />
+    <%= f.text_field :alias_name, autofocus: true %>
+  </div>
+  <!-- 「alias_name」というラベル名はユーザーの混乱を招きそう「ニックネーム」のようなニュアンスで表示させたい」 -->
```
FIXMEなどのアノテーションもコメントしておきましょう

daisakuhazui replied:
了解しました！

ArakawaShintaro commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,11 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  before_action :configure_permitted_parameters, if: :devise_controller?
```
ご自身で調べられたと思うのですが、goodです！

daisakuhazui replied:
ありがとうございます！

Commit:
不要なファイルが自動生成されないようにする

rails g実行時に使用しないファイルが生成されないようにしました。

Commit:
不要なルーティングの削除

static_pagesに関するルーティングを削除しました

Commit:
static_pages_controllerの削除

static_pages_controllerと共に、関連するファイルも削除しました。

Commit:
アノテーションの追加

後々、検索しやすいようにFIXMEを追記しました。

ArakawaShintaro commented:
トップページのUserControllerは作らないのですか？
Userのshowがトップページだと思っていましたが、どうでしょう？

ArakawaShintaro commented:
questionコントローラーindex
userコントローラー show
がトップページになるのでは？

ArakawaShintaro commented:
sign up
sign in
logout
はこのPRで作っちゃいましょう

ArakawaShintaro commented:
github test

Commit:
QuestionContorollerの作成

コントローラの作成と同時にnewアクションとindexアクションも定義しました。

Commit:
アノテーションの追加

余力があれば実装する仕様をメモ書きしました。

Commit:
サインイン画面下部のメッセージを編集

ただ「Sing up」とだけ書かれていたので、サインアップ画面へ誘導しやすいようにメッセージを変更しました。

Commit:
ヘッダーにメニューを追加

サインイン後のユーザーがログアウトできるよう、ヘッダーにメニューを追加しました。
サインインしていないユーザーに対しては、サインイン画面へのリンクメニューを表示させます。

Commit:
ルーティングの追加

rootにはQuestionControllerのindexアクションを設定しました。
また、同じくnewアクションとindexアクションのルーティングも追加しました。

Commit:
QuesitionControllerのビューを作成

newアクションとindexアクションに対応するビューを作成しました。

ArakawaShintaro commented on app/controllers/question_controller.rb:
```diff
@@ -0,0 +1,7 @@
+class QuestionController < ApplicationController
+  def new
```
基本的にindex newの順番が多いですね
scafforldとかで自動生成された時もそうなってますよ

ArakawaShintaro commented on app/views/layouts/application.html.erb:
```diff
@@ -9,6 +9,35 @@
+            <!-- ドロップダウンにする場所ここから -->
+            <li><%=link_to "Profile", '#'%></li>
+            <li><%=link_to "Setting", '#'%></li>
+            <li><%=link_to "Sign out", destroy_user_session_path, method: "delete"%></li>
```
```
<%= link_to "Sign out", destroy_user_session_path, method: "delete" %>
```
前後に半角スペースを開けたいですね

ArakawaShintaro commented on app/views/layouts/application.html.erb:
```diff
@@ -9,6 +9,35 @@
+            <li><%= link_to "Home", '#'%></li>
+
+            <% if signed_in? %>
+            <li><%= link_to "Question", '#'%></li>
```
if文の中はインデントをもう１つ下げたいですね

```
if hoge
  fuga
end
```

ArakawaShintaro replied:
questionに関しては、もうパスが生成されているので#じゃなくてもよくないですか？

ArakawaShintaro commented on config/routes.rb:
```diff
@@ -1,3 +1,6 @@
 Rails.application.routes.draw do
+  root to: 'question#index'
+  get 'question/new'
```
resourcesメソッドを使用しましょう

Commit:
if文内のインデントを修正

一段下げて見やすくしました。

Commit:
アクションを記述する順番を修正

new, indexの順番に記述しました。

Commit:
link_to部分に半角スペースを追加

コードの見た目を整えるために半角スペースを空けました。

Commit:
resourcesメソッドの追加

7つのアクションのルーティングをresourcesで1行で記述しました。

Commit:
ヘッダーメニューへリンクの追加

Quesionsメニューにリンクを追加しました。
遷移先のshowアクションのビューは別のブランチで作成します。

ArakawaShintaro commented on app/controllers/question_controller.rb:
```diff
@@ -1,7 +1,8 @@
-  def index
+  def new
   end
+
```
無駄な改行ですね

ArakawaShintaro commented on config/routes.rb:
```diff
@@ -1,6 +1,5 @@
 Rails.application.routes.draw do
+  resources :questions
```
今回はindexとnewだけなのでそれに制限しましょう

ArakawaShintaro commented on app/controllers/application_controller.rb:
```diff
@@ -1,3 +1,11 @@
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:alias_name, :name])
+  end
+
```
こちらも無駄な改行ですね。。

ArakawaShintaro commented on app/controllers/question_controller.rb:
```diff
@@ -1,7 +1,8 @@
-  def index
+  def new
   end
+
```
そもそもなんですが、サインインの機能でなんでnewってあるんですか！？

daisakuhazui replied:
次のブランチで質問の投稿機能を作ろうとしていたためです。

ただ、この次はユーザープロフィールの登録・編集を実装するべきではないのかと今は考えています。

「ついで」に作ったnewはどの道サインイン機能と無関係ですね・・・！！

ArakawaShintaro commented on app/controllers/question_controller.rb:
```diff
@@ -1,7 +1,8 @@
-  def index
+  def new
   end
+
```
また、無駄な改行が入っています

daisakuhazui replied:
個人的に見やすいと思って改行していました。

ただ、下記のような記事を何件も見つけたので、きちんと読みやすいコードを意識していきます・・・！！
https://teratail.com/questions/73738
https://www.slideshare.net/rootmoon/7-37892729

Commit:
空行の削除

無駄な改行をなくしました。

Commit:
空行の削除

無駄な改行をなくしました。

Commit:
空行の削除とnewアクションの削除

無駄な改行をなくしました。
また、サインアップ機能にnewアクションは今回のブランチで不必要なため削除しました。

Commit:
ルーティングの調整

resourcesメソッドにオプションを追加しました。
サインアップ機能では、サインアップ後にquestionコントローラのindexメソッドのビューに遷移します。
なので今回はindexにのみ限定してあります。

ArakawaShintaro reviewed (commented):
見ました。
ひとまず大丈夫かと！
サインインとログアウトもここで実装しちゃいますか？
